spring-batch-infrastructure/src/main/java/org/springframework/batch/support/transaction/TransactionAwareBufferedWriter.java

```diff
@@ -72,6 +99,47 @@ public void setEncoding(String encoding) {
 		this.encoding = encoding;
 	}
 
+	/**
+	 * Buffers the output on a {@code byte[]} level. Compared to
+	 * buffering on a {@code char[]} level this saves about 50% memory
+	 * for common encodings and input on Java 8.
+	 */
+	static final class Buffer {
+
+		private ByteArrayOutputStream outputStream;
+
+		private OutputStreamWriter writer;
+
+		Buffer(String encoding) throws UnsupportedEncodingException {
+			this.outputStream = new ByteArrayOutputStream();
+			this.writer = new OutputStreamWriter(outputStream, encoding);
+		}
+
+		int length() throws IOException {
```

Comment:
This method is used in `org.springframework.batch.support.transaction.TransactionAwareBufferedWriter#getBufferSize`. According to its Javadoc, the method should return the size of unflushed buffered data. However, the method `length` flushes the writer before returning the size of the outputStream. Do we need to flush the writer to get the length of the buffer? A `get` method should be side-effect free IMO. I tried to remove the `writer.flush` but this makes several tests to fail. Wdyt?

Reply:
The issue is that `OutputStreamWriter` has an internal buffer in `sun.nio.cs.StreamEncoder`. If we want to know how many bytes we will write this will be the bytes in the `ByteArrayOutputStream` plus the bytes in the `OutputStreamWriter` / `sun.nio.cs.StreamEncoder`. There is no way of accessing the number of bytes in the internal buffer in `OutputStreamWriter` / `sun.nio.cs.StreamEncoder`. The only way to be sure is to flush those bytes to the `ByteArrayOutputStream`.